### PR TITLE
feat(medusa,orchestration): Decouple Product in Cart domain

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**

### CODE REVIEW

**Comment by adrien2p** [comment]
Info: 

to reduce the amount of changes required to decouple the product from the cart, the choice has been to not remove the variant id from the line item, and keep the relation `items.variant.product.profiles`. The latter relation is striped out from the retrieve* methods of the cart service. to not break to many tests

File: packages/medusa/src/services/totals.ts [hunk 1]
```diff
@@ -622,6 +622,7 @@ class TotalsService extends TransactionBaseService {
    * @param value - discount value
    * @param discountType - the type of discount (fixed or percentage)
    * @return triples of lineitem, variant and applied discount
+   * @deprecated - Will be removed in a next release
```
Comment (olivermrbl) [inline packages/medusa/src/services/totals.ts]: suggestion: Add "in favor of [new function]" to provide a good experience

File: packages/medusa/src/services/tax-provider.ts [hunk 2]
```diff
@@ -247,46 +247,39 @@ class TaxProviderService extends TransactionBaseService {
...
-          )
-        }
-
-        if (l.variant?.product_id) {
-          return {
-            item: l,
-            rates: productRatesMap.get(l.variant.product_id) ?? [],
-          }
-        }
+    const calculationLines = lineItems.map((item) => {
```
Comment (olivermrbl) [inline packages/medusa/src/services/tax-provider.ts:261]: nice catch

File: packages/medusa/src/services/line-item-adjustment.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ type LineItemAdjustmentServiceProps = {
 }
 
 type AdjustmentContext = {
-  variant: { product_id: string }
+  product_id: string
```
Comment (olivermrbl) [inline packages/medusa/src/services/line-item-adjustment.ts]: suggestion: support both for backward compatibility purposes

File: packages/medusa/src/services/line-item.ts [hunk 4]
```diff
@@ -298,7 +299,9 @@ class LineItemService extends TransactionBaseService {
           if (resolvedContext.cart) {
             const adjustments = await this.lineItemAdjustmentService_
               .withTransaction(transactionManager)
-              .generateAdjustments(resolvedContext.cart, lineItem, { variant })
+              .generateAdjustments(resolvedContext.cart, lineItem, {
+                product_id: lineItem.variant?.product_id,
```
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts]: see other comment

File: packages/medusa/src/services/cart.ts [hunk 5]
```diff
@@ -2163,10 +2188,31 @@ class CartService extends TransactionBaseService {
...
+          if (
+            this.featureFlagRouter_.isFeatureEnabled(
+              IsolateProductDomainFeatureFlag.key
+            )
+          ) {
+            productShippinProfileMap =
+              await this.shippingProfileService_.getMapProfileIdsByProductIds(
+                cart.items.map((item) => item.variant.product_id)
+              )
+          }
```
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts]: suggestion: Can we save a mapping here?
```suggestion
          let productShippinProfileMap = new Map<string, string>()

          if (
            this.featureFlagRouter_.isFeatureEnabled(
              IsolateProductDomainFeatureFlag.key
            )
          ) {
            productShippinProfileMap =
              await this.shippingProfileService_.getMapProfileIdsByProductIds(
                cart.items.map((item) => item.variant.product_id)
              )
          } else {
            productShippinProfileMap = cart.items.map((item) => [
              item.variant?.product?.id,
              item.variant?.product?.profile_id,
            ])
          }
```

File: packages/medusa/src/services/line-item-adjustment.ts (hunk 3 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/line-item-adjustment.ts]: reverted it to the original object as the variant.product_id is being populated by a typeorm hook

**Review by olivermrbl — APPROVED** [review]

**Pull request merged**